botocore/endpoints_v2.py

```diff
@@ -0,0 +1,1075 @@
+                }
+                },
+                {
+                "type": "error",
+                "conditions": [
+                    // Abbreviated for clarity
+                ],
+                "error": "an error message"
+                }
+            ]
+            }
+        ]
+    }
+    """
+
+    def __init__(self, rules, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.rules = [
+            RuleCreator.create(partitions=self.partitions, **rule)
+            for rule in rules
+        ]
+
+    def evaluate(self, input_parameters):
+        """If a tree rule's conditions evaluate successfully, iterate over its
+        subordinate rules and return a result if there is one. If any of the
+        subsequent rules are trees, the function will recurse until it reaches
+        an error or an endpoint rule.
+
+        :returns: An endpoint or error object.
+        """
+        if self.evaluate_conditions(input_parameters) is True:
+            for rule in self.rules:
+                # newly set parameters via 'assign' cannot be shared between
+                # adjacent rules
+                evaluation = rule.evaluate(input_parameters.copy())
+                if evaluation:
+                    return evaluation
+        return None
+
+
+class RuleCreator:
+    """A helper class that creates rule objects."""
+
+    endpoint = EndpointRule
+    error = ErrorRule
+    tree = TreeRule
+
+    @classmethod
+    def create(cls, **kwargs):
+        """Given a rule dictionary, locate the `type` attribute, using it to
+        determine the type of rule object to create. Then create and return it.
+
+        :type kwargs: dict
+        :params kwargs: An arbitrary dictionary containing required parameters
+        for different types of rule objects i.e. conditions, rules, errors,
+        endpoints etc.
+
+        :returns: A TreeRule/EndpointRule/ErrorRule object
+        """
+        rule_type = kwargs.pop('type')
+        try:
+            rule_klass = getattr(cls, rule_type)
+        except AttributeError:
+            raise RuleSetConfigurationError(
+                msg=f"Unknown rule type: {rule_type}. A rule must "
+                "be of type 'tree', 'endpoint' or 'error'."
+            )
+        else:
+            return rule_klass(**kwargs)
+
+
+class ParameterDefinition:
+    """A class that houses the spec of an individual parameter defined in a
+    rule set object i.e.
+    {
+        "Region": {
+            "builtIn": "AWS::Region",
+            "required": true,
+            "type": "String"
+        }
+    }
+    """
+
+    def __init__(
+        self,
+        name,
+        parameter_type,
+        documentation=None,
+        builtIn=None,
+        default=None,
+        required=None,
+        deprecated=None,
+    ):
+
+        self.name = name
+        self.parameter_type = getattr(
+            self.ParameterType, parameter_type.lower()
+        ).value
```

Comment:
What's our failure state if we get a new ruleset with a different type? This is going to throw an AttributeError I believe but what bubbles up to the end user?